### PR TITLE
LibWeb/HTML: Resolve image decoding promises inside tasks

### PR DESCRIPTION
**Pull request by AtkinsSJ**
This corresponds to https://github.com/whatwg/html/pull/10753

While I was here, I've also implemented the check for the current request changing, and reformatted the spec comments to match our style.

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp [hunk 1]
```diff
@@ -311,55 +314,97 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<WebIDL::Promise>> HTMLImageElement::decode(
...
+        // 3. Otherwise, in parallel wait for one of the following cases to occur, and perform the corresponding actions:
+        Platform::EventLoopPlugin::the().deferred_invoke(JS::create_heap_function(heap(), [this, promise, &realm, &global] {
+            enum class ImageState {
+                DocumentNoLongerFullyActive,
+                CurrentRequestChanged,
+                CurrentRequestBroken,
+                CurrentRequestCompletelyAvailable,
+            };
+            ImageState image_state;
+            auto previous_state = current_request().state();
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp:334]: This double event loop spin has always been truly puzzling. ... why do we need it?

File: Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp [hunk 2]
```diff
@@ -311,55 +314,97 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<WebIDL::Promise>> HTMLImageElement::decode(
...
-        Platform::EventLoopPlugin::the().deferred_invoke(JS::create_heap_function(heap(), [this, promise, &realm, reject_if_document_not_fully_active, reject_if_current_request_state_broken] {
+        // 3. Otherwise, in parallel wait for one of the following cases to occur, and perform the corresponding actions:
+        Platform::EventLoopPlugin::the().deferred_invoke(JS::create_heap_function(heap(), [this, promise, &realm, &global] {
+            enum class ImageState {
+                DocumentNoLongerFullyActive,
+                CurrentRequestChanged,
+                CurrentRequestBroken,
+                CurrentRequestCompletelyAvailable,
+            };
+            ImageState image_state;
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp:333]: :/ not initialized at declaration.

File: Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp [hunk 3]
```diff
@@ -311,55 +314,97 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<WebIDL::Promise>> HTMLImageElement::decode(
...
+            auto previous_state = current_request().state();
             Platform::EventLoopPlugin::the().spin_until(JS::create_heap_function(heap(), [&] {
                 auto state = this->current_request().state();
 
-                return !this->document().is_fully_active() || state == ImageRequest::State::Broken || state == ImageRequest::State::CompletelyAvailable;
-            }));
+                if (state != previous_state) {
+                    image_state = ImageState::CurrentRequestChanged;
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp:339]: does the state of the current request changing really mean that the current request changed? What about ABA? Should we be storing the "Current request" and making sure the image element still has the same request instead?

Comment (ADKaster) [reply Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp:339]: "or is mutated" :thinking: How do we detect that?

File: Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp (hunk 1 above)
Comment (AtkinsSJ) [reply Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp:334]: I'm confused, I only see one spin here.

**Comment by AtkinsSJ** [comment]
Changes:

- Replaced the "spin and set an enum, then act on the enum" code with "just do the action in the spin callback".
- Removed the "request changed or was mutated" check for now, because it's clear I was not actually doing that correctly before, and it's not at all obvious what the correct thing is.

**Comment by tcl3** [comment]
This has conflicts.

**Comment by AtkinsSJ** [comment]
Fixed the conflict and updated to use the new LibGC renamed methods. (I may have forgotten about this PR... 😅)

**Pull request merged**